### ISSUE TITLE
Sitemap generation

### ISSUE THREAD

**Issue by synalysis** [dillonkearns/elm-pages#31]
In order to be SEO friendly sitemap generation would be a great feature.

**Comment by dillonkearns** [comment]
Hey @synalysis, I  completely agree! This has been something I've been thinking about.

I think what it will look like in `elm-pages` is first to have a "hook" in your application config (a function that you pass in) that allows you to generate arbitrary files based on all of the metadata for your site.

Something like:

```elm
type alias FileToGenerate = { name : String, content : String }

Pages.Platform.application
  {
    generateFiles : List { metadata : Metadata, path : PagePath Pages.PathKey } -> List FileToGenerate
    -- ...
}
```

Then from there, users can build their own sitemap extensions easily.

**Comment by synalysis** [comment]
This sounds like a promising idea - very generic and extensible. And with the metadata it would be possible to exclude certain pages like drafts from the sitemap.

For elm-pages-starter sitemap generation could be then a sample implementation.

Do you see chances to pass the last modification date of the content .md files somehow? This could be used for the <lastmod>2019-12-12T14:12:33+01:00</lastmod> entry.

**Comment by dillonkearns** [comment]
Yeah, I've been thinking about adding some metadata like the last modified time of files in the content folder. I think that would definitely be a powerful and sensible feature. It might be as simple as generating that as part of the generated files. I'll have to play around with that idea a little more.

By the way, for reference (for my future self, or anyone else who likes reading through specs), here's a spec for sitemaps: https://www.sitemaps.org/protocol.html. I always find it nice to browse through those when designing features.

I totally agree, doing this for `elm-pages-starter` would be a great way to develop the design and make sure it works well.

**Comment by dillonkearns** [comment]
I just released Elm package version 2.0.0 and NPM version 2.0.0 with a `generateFiles` hook. I also released two small packages, `dillonkearns/elm-rss` and `dillonkearns/elm-sitemap`. I wired them into the elm-pages-starter, here's the commit:

https://github.com/dillonkearns/elm-pages-starter/commit/5d4570456d483b925c7cfd2fbb7c770efd018778

I hope this opens up a lot of new possibilities with elm-pages!
